drivers/pinctrl/pinctrl_ite_it8xxx2.c

```diff
@@ -89,23 +118,110 @@ static int pinctrl_it8xxx2_set(const pinctrl_soc_pin_t *pins)
+	 * If KSI[7:0]/KSO[15:0] is in KBS mode , setting 1 enables the internal
+	 * pull-up (KSO[17:16] setting internal pull-up by GPIO port GPCR register).
+	 * If KSI[7:0]/KSO[15:0] is in GPIO mode, then this bit is always disabled.
+	 */
+	switch (IT8XXX2_DT_PINCFG_PULLUP(pincfg)) {
+	case IT8XXX2_PULL_PIN_DEFAULT:
+		/* Disable internal pulll-up */
+		*reg_ctrl &= ~pullup_mask;
+		break;
+	case IT8XXX2_PULL_UP:
+		*reg_ctrl |= pullup_mask;
+		break;
+	default:
+		LOG_ERR("This pull level is not supported.");
+		return -EINVAL;
+	}
+
+	/*
+	 * Set push-pull or open-drain mode (this bit apply to all pins):
+	 * KSI[7:0] doesn't support push-pull and open-drain settings in kbs mode.
+	 * If KSO[17:0] is in KBS mode, setting 1 selects open-drain mode,
+	 * setting 0 selects push-pull mode.
+	 * If KSO[15:0] is in GPIO mode, then this bit is always disabled.
+	 */
+	if (pp_od_mask != NO_FUNC) {
+		switch (IT8XXX2_DT_PINCFG_PP_OD(pincfg)) {
+		case IT8XXX2_PUSH_PULL:
+			*reg_ctrl &= ~pp_od_mask;
+			break;
+		case IT8XXX2_OPEN_DRAIN:
+			*reg_ctrl |= pp_od_mask;
+			break;
+		default:
+			LOG_ERR("This pull mode is not supported.");
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static int pinctrl_kscan_it8xxx2_configure_pins(const pinctrl_soc_pin_t *pins,
+						uint8_t pin_cnt, uint8_t i)
+{
+	const struct pinctrl_it8xxx2_config *pinctrl_config = pins[i].pinctrls->config;
+	const struct pinctrl_it8xxx2_ksi_kso *ksi_kso = &(pinctrl_config->config.ksi_kso);
+	volatile uint8_t *reg_gctrl = ksi_kso->reg_gctrl;
+	uint8_t pin_mask = BIT(pins[i].pin);
+
+	/* Set a pin of KSI[7:0]/KSO[15:0] to pullup, push-pull/open-drain */
+	if (pinctrl_kscan_it8xxx2_set(&pins[i])) {
+		return -EINVAL;
+	}
+
+	/* Set a pin of KSI[7:0]/KSO[15:0] to kbs mode */
+	*reg_gctrl &= ~pin_mask;
+
+	return 0;
+}
+
 int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
 			   uintptr_t reg)
 {
 	ARG_UNUSED(reg);
 
 	const struct pinctrl_it8xxx2_config *pinctrl_config;
+	const struct pinctrl_it8xxx2_gpio *gpio;
 	volatile uint8_t *reg_gpcr;
 	volatile uint8_t *reg_func3_gcr;
 	volatile uint8_t *reg_func4_gcr;
 	uint8_t pin;
+	int status;
 
 	for (uint8_t i = 0U; i < pin_cnt; i++) {
 		pinctrl_config = pins[i].pinctrls->config;
 		pin = pins[i].pin;
-		reg_gpcr = (uint8_t *)pinctrl_config->reg_gpcr + pin;
-		reg_func3_gcr = (uint8_t *)(pinctrl_config->func3_gcr[pin]);
-		reg_func4_gcr = (uint8_t *)(pinctrl_config->func4_gcr[pin]);
+
+		/* Pinctrl for KSI[7:0] and KSO[15:0] */
+		if (!(pinctrl_config->gpio_group)) {
+			status = pinctrl_kscan_it8xxx2_configure_pins(pins, pin_cnt, i);
+			if (status < 0) {
+				const char *name = pins[i].pinctrls->name;
+
+				LOG_ERR("%s pin%d configuration is invalid.", name, pin);
+				return status;
+			}
+			continue;
+		}
+
+		gpio = &(pinctrl_config->config.gpio);
+		reg_gpcr = (uint8_t *)gpio->reg_gpcr + pin;
+		reg_func3_gcr = (uint8_t *)(gpio->func3_gcr[pin]);
+		reg_func4_gcr = (uint8_t *)(gpio->func4_gcr[pin]);
```

Comment:
I recommend you create a `pinctrl_gpio_it8xxx_configure_pins()` routine.  Then the public `pinctrl_configure_pins()` becomes:

```
int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
			   uintptr_t reg)
	for (uint8_t i = 0U; i < pin_cnt; i++) {
		if (pins[i].pinctrls->config->gpio_group) {
			status = pinctrl_gpio_it8xxx2_configure_pins(&pins[i]);
		} else {
			status = pinctrl_kscan_it8xxx2_configure_pins(&pins[i]);
		}
		if (status < 0) {
			LOG_ERR("...");
			return status;
		}
	}
	return 0;
}
```

Reply:
Done